core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -289,13 +290,26 @@ private[spark] class ExecutorAllocationManager(
       s" tasksperexecutor: $tasksPerExecutor")
     val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
       tasksPerExecutor).toInt
-    if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
+    val maxNeededWithSpeculationLocalityOffset =
+      if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
       // If we have pending speculative tasks and only need a single executor, allocate one more
       // to satisfy the locality requirements of speculation
       maxNeeded + 1
     } else {
       maxNeeded
     }
+
+    // Since the maxNeededWithSpeculationLocalityOffset already includes the num executors needed
+    // to run unschedulable tasks, we would only try to add when the
+    // maxNeededWithSpeculationLocalityOffset is lesser than the active executors
```

Comment:
this comment is also confusing to me.  its also above a check for numUnschedulables, which has nothing to with the comment.  Can we move it down next to the check on line 308 and clarify. I think you mean we don't add any tasks for the unschedulables if active executors is less than the number we are already asking for.